File: analysis/orcamento_funcional.py
import pandas as pd
from sqlalchemy import text


def get_orcamento_funcional(conn, year: int, empresa_ids: list[str] | None = None) -> pd.DataFrame:
    empresa_clause = "AND empresa_id = ANY(:empresas)" if empresa_ids else ""
    params: dict = {"ano": year}
    if empresa_ids:
        params["empresas"] = empresa_ids
    query = text(f"""
        SELECT funcao_nome, dotacao_atualizada, subfuncao_nome, empenhado, liquidado, pago
        FROM fct_despesas
        WHERE ano = :ano AND funcao_nome IS NOT NULL {empresa_clause}
    """)
    df = pd.read_sql_query(query, conn, params=params)

    # Converte colunas monetárias em texto para numérico
    for col in ["dotacao_atualizada", "empenhado", "liquidado", "pago"]:
        df[col] = pd.to_numeric(df[col].astype(str).str.replace(",", "."), errors="coerce").fillna(0)

    return df.groupby(["funcao_nome", "subfuncao_nome"], as_index=False).agg(
        dotacao_atualizada=("dotacao_atualizada", "sum"),
        empenhado=("empenhado", "sum"),
        liquidado=("liquidado", "sum"),
        pago=("pago", "sum"),
    )

File: packages/db/src/queries/orcamento_funcional.ts
import { sql } from "kysely";
import { db } from "../client";

export interface ItemOrcamentoFuncional {
  funcao_nome: string;
  subfuncao_nome: string;
  dotacao_atualizada: number;
  empenhado: number;
  liquidado: number;
  pago: number;
}

export async function getOrcamentoFuncional(
  year: number,
  empresaIds?: string[] | null
): Promise<ItemOrcamentoFuncional[]> {
  try {
    let q = sql`
      SELECT funcao_nome, subfuncao_nome, dotacao_atualizada, empenhado, liquidado, pago
      FROM fct_despesas
      WHERE ano = ${year} AND funcao_nome IS NOT NULL
    `;
    if (empresaIds && empresaIds.length > 0) {
      q = sql`${q} AND empresa_id = ANY(${empresaIds})`;
    }
    const res = await q.execute(db);
    const rows = (res.rows as any[]) || [];

    const map: Record<
      string,
      {
        funcao_nome: string;
        subfuncao_nome: string;
        dotacao_atualizada: number;
        empenhado: number;
        liquidado: number;
        pago: number;
      }
    > = {};

    for (const r of rows) {
      const func = String(r.funcao_nome ?? "");
      const sub = String(r.subfuncao_nome ?? "");
      const key = `${func}__${sub}`;

      const dot = parseFloat(String(r.dotacao_atualizada ?? "0").replace(",", ".")) || 0;
      const emp = parseFloat(String(r.empenhado ?? "0").replace(",", ".")) || 0;
      const liq = parseFloat(String(r.liquidado ?? "0").replace(",", ".")) || 0;
      const pag = parseFloat(String(r.pago ?? "0").replace(",", ".")) || 0;

      if (!map[key]) {
        map[key] = {
          funcao_nome: func,
          subfuncao_nome: sub,
          dotacao_atualizada: 0,
          empenhado: 0,
          liquidado: 0,
          pago: 0,
        };
      }
      map[key].dotacao_atualizada += dot;
      map[key].empenhado += emp;
      map[key].liquidado += liq;
      map[key].pago += pag;
    }

    return Object.values(map);
  } catch {
    return [];
  }
}
